### PR TITLE
Add support for querying ADXL345 accelerometers via SPI

### PR DESCRIPTION
**Pull request by KevinOConnor**
This adds support for querying an ADXL345 accelerometer using Klipper's micro-controller SPI interface.  It's been in discussion in issue #3027 .  This has been tested with an ADXL345 wired to a Raspberry Pi that has been setup to use Klipper's "linux mcu" support.  However, it should work on any micro-controller that Klipper supports.

This PR only adds support for bulk reading of accelerometer data.  It is expected that resonance testing and analysis will be added at a later date.

The current code adds an ACCELEROMETER_MEASURE command.  However, I don't much like the current command name, current command format, nor resulting file format.  I don't have any better ideas right now.  I expect we'll tweak the command and formats prior to merging.

To test this setup:
`cd ~/klipper ; git fetch ; git checkout origin/work-adxl345-20200730 ; sudo service klipper stop ; make flash ; sudo service klipper start`

On an RPi, see the wiring guide described at https://github.com/nagimov/adxl345spi#wiring and add the following to the printer.cfg file:
```
[mcu rpi]
serial: /tmp/klipper_host_mcu

[adxl345 my_accel]
cs_pin: rpi:None
```

-Kevin

### CODE REVIEW

**Comment by KevinOConnor** [comment]
@dmbutyugin - FYI, I attempted to merge your resonance_testing.py module into this branch.  It looks like it is running okay.

-Kevin

**Comment by dmbutyugin** [comment]
@KevinOConnor that's great, I think there is no reason for the resonance_tester not to work. BTW, I think we could get rid of the heuristic to find the beginning of the move if we change the C API to measure the data from a certain point in time, and not from 'now'. Then, if that's possible, we could schedule the measurements of acceleration from exactly the point in time when the toolhead starts a move.

**Comment by KevinOConnor** [comment]
> BTW, I think we could get rid of the heuristic to find the beginning of the move if we change the C API to measure the data from a certain point in time, and not from 'now'. Then, if that's possible, we could schedule the measurements of acceleration from exactly the point in time when the toolhead starts a move.

FYI, the micro-controller code does schedule the start at a requested time.  Alas, it's not easy to predict when an upcoming move will occur after a complete move queue flush though.  (The host processing time that it takes to fill the move queue will skew when the first move in the queue gets scheduled to run at.)  I think it should be possible to figure out when that first move was actually started though, and then use that time to prune out the pre-move accelerometer data..

Separate, a couple of things in the resonance_tester.py code that we might want to discuss:
1. The code should probably use `toolhead.move()` instead of `gcode.cmd_G1()`.  Otherwise, if the user has bed_mesh (or bed_tilt) defined then it could skew the resonance test.  Also, `toolhead.move()` is faster.
2. I think the TEST_VIBRATIONS command may be "over emphasizing" low frequency (10-25hz) resonance.  I've seen a bunch of graphs of TEST_VIBRATIONS output that'll show similar peaks at, for example,10hz and 50hz.  But I suspect a 10x of 10hz accel isn't nearly as much a problem as 10x of 50hz accel.  (Also, low frequency vibration may be the whole printer shaking, which may not be a problem as the toolhead and bed may actually be vibrating in sync.) So, the graph may be misleading at low frequencies.  That said, I don't have an alternative.

Cheers,
-Kevin

**Comment by KevinOConnor** [comment]
Oh, and:
3. In TEST_VIBRATIONS, I wonder if it would be better to calculate resonance using the magnitude of the acceleration vector (`sqrt(accel_x**2 + accel_y**2 + accel_z**2)`) instead of calculating each axis independently.  If the accelerometer is not mounted perfectly square, it'll skew the results of each axis, but it shouldn't adversely impact the vector.

-Kevin

**Comment by dmbutyugin** [comment]
@KevinOConnor 
> FYI, the micro-controller code does schedule the start at a requested time. Alas, it's not easy to predict when an upcoming move will occur after a complete move queue flush though.

Yep, that's what I was afraid of. Well, the heuristic work reasonably well (for a test version of the code).


>     1. The code should probably use `toolhead.move()` instead of `gcode.cmd_G1()`.  Otherwise, if the user has bed_mesh (or bed_tilt) defined then it could skew the resonance test.  Also, `toolhead.move()` is faster.

Yes, probably. You have a good point.
 
>     2. I think the TEST_VIBRATIONS command may be "over emphasizing" low frequency (10-25hz) resonance.  I've seen a bunch of graphs of TEST_VIBRATIONS output that'll show similar peaks at, for example,10hz and 50hz.  But I suspect a 10x of 10hz accel isn't nearly as much a problem as 10x of 50hz accel.  (Also, low frequency vibration may be the whole printer shaking, which may not be a problem as the toolhead and bed may actually be vibrating in sync.) So, the graph may be misleading at low frequencies.  That said, I don't have an alternative.

In general, the amplitude of oscillations grows as (1 / freq) or as (1 / freq)^2 (depending on how you estimate), so in general the lower-frequency oscillations are more problematic. However, I agree that if the whole frame is shaking in unison, then it will likely be less of a problem. Also, in my opinion the precision of the measurements below 10-15 Hz is not very high. I have some ideas about different versions of the test, maybe they will work better in the end. I plan to try a few things next weekend.

> 3. In TEST_VIBRATIONS, I wonder if it would be better to calculate resonance using the magnitude of the acceleration vector (`sqrt(accel_x**2 + accel_y**2 + accel_z**2)`) instead of calculating each axis independently. If the accelerometer is not mounted perfectly square, it'll skew the results of each axis, but it shouldn't adversely impact the vector.

Well, right now the code generates all 3 values for each axis. At least in the test code, this can help users find some (hopefully useful) correlations between the axes. Separately, the produced triplets of values are additive, so one could plot 1 chart that shows  val_x + val_y + val_z. Or sqrt(...) of that value.

**Comment by inks007** [comment]
Can confirm that the ADXL345 works off an STM32F205 (toolboard shield feasibility tests). I am testing with 6" jumper wires via RPi4 UART5 @ 250kBaud though.

/tmp/adxl345-20200824_002054.csv contents (90k lines):
```
##drops=16,overflows=0,time_per_sample=0.000305819,start_range=0.000006,end_range=0.000006
#time,accel_x,accel_y,accel_z
206.880614,9571.290400,-3844.206800,-1098.344800
206.880920,9414.384000,-4393.379200,-1961.330000
206.881226,9414.384000,-4314.926000,-2039.783200
...
```
Doesn't look like it uses the two least significant digits.

`Stats 191845.0: gcodein=302 mcu: mcu_awake=0.000 mcu_task_avg=0.000003 mcu_task_stddev=0.000003 bytes_write=6529 bytes_read=784714 bytes_retransmit=0 bytes_invalid=0 send_seq=1031 receive_seq=1031 retransmit_seq=0 srtt=0.001 rttvar=0.000 rto=0.025 ready_bytes=0 stalled_bytes=0 freq=120000007 print_time=836.137 buffer_time=0.000 print_stall=0 sysload=0.00 cputime=11.554 memavail=7610200`

**Comment by KevinOConnor** [comment]
> Can confirm that the ADXL345 works off an STM32F205 (toolboard shield feasibility tests).

Thanks.  I also got word the other day that one tester had success reading from an adxl345 via an atmega328p.  (If anyone tries with an AVR chip, be sure not to mix 5V and 3.3V on the adxl345.)

> ##drops=16,overflows=0,time_per_sample=0.000305819,start_range=0.000006,end_range=0.000006

The `drops=16` is not a good sign as it indicates lost messages between host and micro-controller.  I suppose if you got 90k results and only lost 16 messages then that isn't too bad.  Ideally there would be zero drops.  Almost assuredly the reason for the drops is because you are using 250000 baud.  At that baud the chip can send 25,000 bytes per second and measurements will require ~23200 bytes per second.  In contrast, bandwidth should not be a problem if using any of the chips with native USB support.

> Doesn't look like it uses the two least significant digits.

Yes, that's normal.

-Kevin

**Comment by randellhodges** [comment]
Not sure if this an appropriate place to add a mini feature request, but, there are situations where the board is mounted in a way such that a ADXL345 axis does not match the printers.

Would it be possible to add a configuration option under the `[adxl345]` section to map an ADX345 axis to the printers axis? That way the results match up?

**Comment by KevinOConnor** [comment]
> Would it be possible to add a configuration option under the [adxl345] section to map an ADX345 axis to the printers axis?

Yes, that's one of the things still needed to be implemented.

-Kevin

**Comment by KevinOConnor** [comment]
FYI, this branch has been superseded by @dmbutyugin's branch at: https://github.com/dmbutyugin/klipper/tree/work-adxl345-20200730

@dmbutyugin - my plan right now is to fix up a few minor things in the low-level adxl345.py code and then merge that into the master branch.  We can then work on merging your resonance_tester.py code on top.  Let me know if you have any concerns with that approach.

-Kevin

**Comment by dmbutyugin** [comment]
@KevinOConnor, that sounds good. I will have to do some refactoring of adxl345.py for remote process(ing) too, but I think it can be done as a part of the integration of the shaper calibration code.

BTW, since I've been running resonance testing pretty extensively lately, I noticed that I quite frequently get some small number of reported overflows (of FIFO queue), e.g. just now
```
ADXL345 stats: drops=0,overflows=1,time_per_sample=0.000322696,start_range=0.000063,end_range=0.000033
```
If happens, usually overflows are within 1-3 range. I cannot say it happens very frequently though, but it is not a rare event either. Do you have any thoughts on this? I'm running Klipper on Raspberry Pi 3 B+, and the accelerometer is connected to the same Pi via MCU host.

**Comment by KevinOConnor** [comment]
> BTW, since I've been running resonance testing pretty extensively lately, I noticed that I quite frequently get some small number of reported overflows (of FIFO queue)

I've seen this as well.  One problem was that if a previous set of measurements wasn't shutdown properly, then it'd cause an overflow at the start of the next measurement.  That should be fixed now.

I guess it's also possible that an spi read error occurred, or perhaps even the Linux kernel didn't schedule the mcu code in sufficient time.

-Kevin

**Comment by KevinOConnor** [comment]
FYI, I've rebase'd this code.  Some new features:
1. It's no longer necessary to specify a chip name in the adxl345 config.  That is, most users can just use `[adxl345]`.
2. It's now possible to specify an `axes_map` parameter to remap the accelerometer axes if they are not aligned with the printer axes.
3. It's possible to specify a NAME parameter to ACCELEROMETER_MEASURE which alters the ultimate filename that results are written to.
4. Some minor low-level adxl345 setup fixes.

This branch is still superseded by the latest resonance testing methods at: https://github.com/dmbutyugin/klipper/tree/work-adxl345-20200730

My goal is to merge just the first two patches of this series (which contain the low-level adxl345 chip communication code).

-Kevin

File: docs/G-Codes.md [hunk 1]
```diff
@@ -633,3 +633,17 @@ section is enabled:
...
+The following command is available when an "adxl345" config section is
+enabled:
+- `ACCELEROMETER_MEASURE [CHIP=<config_name>] [RATE=<value>]
+  [NAME=<value>]`: Starts accelerometer measurements at the requested
+  number of samples per second. If CHIP is not specified it defaults
+  to "default". Valid rates are 25, 50, 100, 200, 400, 800, 1600,
+  and 3200. If RATE is zero (or not specified) then the current series
+  of measurements are stopped and the results are written to a file
+  named `/tmp/adxl345-<name>.csv` where "<name>" is the optional NAME
+  parameter. If NAME is not specified it defaults to the current time
```
Comment (Sineos) [inline docs/G-Codes.md:648]: > If RATE is zero (or not specified) then the current series of measurements are [...]

I find this a little bit confusing. Isn't the RATE 3200 if omitted?

**Pull request merged**